smack-extensions/src/main/java/org/jivesoftware/smackx/mood/MoodManager.java

```diff
@@ -0,0 +1,177 @@
+import org.jivesoftware.smack.SmackException;
+import org.jivesoftware.smack.XMPPConnection;
+import org.jivesoftware.smack.XMPPException;
+import org.jivesoftware.smack.packet.Message;
+import org.jivesoftware.smack.provider.ProviderManager;
+import org.jivesoftware.smack.util.Async;
+import org.jivesoftware.smackx.disco.ServiceDiscoveryManager;
+import org.jivesoftware.smackx.mood.element.MoodConcretisation;
+import org.jivesoftware.smackx.mood.element.MoodElement;
+import org.jivesoftware.smackx.mood.provider.MoodConcretisationProvider;
+import org.jivesoftware.smackx.pep.PEPListener;
+import org.jivesoftware.smackx.pep.PEPManager;
+import org.jivesoftware.smackx.pubsub.EventElement;
+import org.jivesoftware.smackx.pubsub.ItemsExtension;
+import org.jivesoftware.smackx.pubsub.LeafNode;
+import org.jivesoftware.smackx.pubsub.PayloadItem;
+import org.jivesoftware.smackx.pubsub.PubSubException;
+import org.jivesoftware.smackx.pubsub.PubSubManager;
+
+import org.jxmpp.jid.BareJid;
+import org.jxmpp.jid.EntityBareJid;
+
+/**
+ * Entry point for Smacks API for XEP-0107: User Mood.
+ *
+ * To set a mood, please use one of the {@link #setMood(Mood)} methods. This will publish the users mood to a pubsub
+ * node.<br>
+ * <br>
+ * In order to get updated about other users moods, register a {@link MoodListener} at
+ * {@link #addMoodListener(MoodListener)}. That listener will get updated about any incoming mood updates of contacts.<br>
+ * <br>
+ * To stop publishing the users mood, refer to {@link #clearMood()}.<br>
+ * <br>
+ * It is also possible to add {@link MoodElement}s to {@link Message}s by using {@link #addMoodToMessage(Message, Mood)}.<br>
+ * <br>
+ * The API can be extended with custom mood concretisations by extending {@link MoodConcretisation} and registering
+ * {@link MoodConcretisationProvider}s using {@link ProviderManager#addExtensionProvider(String, String, Object)}.<br>
+ * An example of how this can be done can be found in the MoodConcretisationTest in the test package.
+ *
+ * @see <a href="https://xmpp.org/extensions/xep-0107.html">
+ *     XEP-0107: User Mood</a>
+ */
+public final class MoodManager extends Manager {
+
+    public static final String MOOD_NODE = "http://jabber.org/protocol/mood";
+    public static final String MOOD_NOTIFY = MOOD_NODE + "+notify";
+
+    private static final Map<XMPPConnection, MoodManager> INSTANCES = new WeakHashMap<>();
+
+    private final Set<MoodListener> moodListeners = new HashSet<>();
+
+    private MoodManager(XMPPConnection connection) {
+        super(connection);
+        PEPManager.getInstanceFor(connection).addPEPListener(new PEPListener() {
+            @Override
+            public void eventReceived(final EntityBareJid from, final EventElement event, Message message) {
+                if (MOOD_NODE.equals(event.getEvent().getNode())) {
+                    final BareJid contact = from.asBareJid();
+                    Async.go(new Runnable() {
```

Comment:
This will lose the order of the mood events. You want to use Smack's new `AsyncButOrdered` tool here.